cosmwasm/packages/wasmi-runtime/src/wasm/contract_operations.rs

```diff
@@ -214,10 +215,59 @@ fn start_engine(
 
     // Create a parity-wasm module first, so we can inject gas metering to it
     // (you need a parity-wasm module to use the pwasm-utils crate)
-    let p_modlue = elements::deserialize_buffer(contract).map_err(|_| EnclaveError::InvalidWasm)?;
+    let mut p_modlue: Module =
+        elements::deserialize_buffer(contract).map_err(|_| EnclaveError::InvalidWasm)?;
 
     trace!("Deserialized Wasm contract");
 
+    let memory_section = p_modlue
+        .memory_section_mut()
+        .ok_or(EnclaveError::CannotInitializeWasmMemory).map_err(|err| {
+            error!(
+                "Error accessing memory section of WASM while trying to validate memory demands: {:?}",
+                err
+            );
+            err
+        })?;
+
+    if memory_section.entries().len() != 1 {
+        error!(
+            "WASM demands too many memory sections. Must be 1, demands {}",
+            memory_section.entries().len()
+        );
+        return Err(EnclaveError::CannotInitializeWasmMemory);
+    }
+
+    let memory_entry = memory_section
+        .entries()
+        .first()
+        .ok_or(EnclaveError::CannotInitializeWasmMemory)
+        .map_err(|err| {
+            error!(
+                "Error accessing memory entry of WASM while trying to validate memory demands: {:?}",
+                err
+            );
+            err
+        })?;
+
+    let requested_initial_pages: u32 = memory_entry.limits().initial();
+    let maximum_allowed_pages: u32 = 192; // 12 MiB
+
+    if requested_initial_pages > maximum_allowed_pages {
+        error!(
+            "WASM Requested to initialize with {} pages, maximum allowed is {}",
+            requested_initial_pages, maximum_allowed_pages
+        );
+        return Err(EnclaveError::CannotInitializeWasmMemory);
+    }
+
+    *memory_section.entries_mut() = vec![MemoryType::new(
+        requested_initial_pages,
+        Some(maximum_allowed_pages),
+    )];
+
+    trace!("Validated WASM memory demands");
+
```

Comment:
This seems like the type of thing that would be cleaner to separate to a different file/class

Reply:
It looks big because all the `map_err` and `error!()` out there, and I really don't like the amount of logic wrapping we are doing as it is, but your call...

Reply:
The point is to reduce the amount of information thrown at someone trying to understand what your code is doing. In this case you're doing memory initialization in the function "start_engine" which isn't trivial. You can move all the memory stuff into a different function at the very least, but for me personally it's not intuitive that memory initialization of the contract is in a file called "contract_operations.rs". Moving this stuff to "memory.rs" I think would make things clearer

Reply:
Done 5b6c07c